Add some endpoints for applications

cderici opened the pull request:
This adds the following endpoints:

- `getApplication`  - `GET /application/{name}` for `Application.ApplicationsInfo()` 
- `addApplication` - `POST /application/{name}?force=false&dry-run=false&trust=false` for `Application.DeployFromRepository()`
- `deleteApplication` - `DELETE /application/{name}` for `Application.DestroyApplication()`
- `refreshApplication` - `PATCH /application/{name}` for `Application.SetCharm()`
- `relateApplication` - `POST /application/relate` for `Application.AddRelation()`

#### Some notes
- Some custom responses are added, e.g., `ApplicationDestroyed`, `RelationCreated`, etc.
- Some schema types are renamed (not the existing ones, but the ones coming from the juju facades), e.g., `ApplicationResult -> Application`, not sure if I should document all the renames.
- Reviewer be advised, I just learned about the openAPI spec this afternoon and this is literally the first time I'm writing one, so there's probably some errors in there.

cderici committed:
Fix some references

cderici committed:
Add some application endpoints

This adds:
- getApplication
- addApplication
- deleteApplication
- refreshApplication
- relateApplication

cderici committed:
Add default values to boolean flags for deploy

cderici committed:
Update the progress table

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
       description: Find out more about Juju clouds.
       url: https://juju.is/docs/olm/manage-clouds
 paths:
+  /application/{name}:
```
I wonder if this should be `/model/{name}/application/{name}`? Imagine you're interacting with this API using `curl` or similar, how would the server know which model you're referring to?

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
       description: Find out more about Juju clouds.
       url: https://juju.is/docs/olm/manage-clouds
```
You need to add a section here to define the `application` tag :)

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+          $ref: "#/components/responses/Forbidden"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+    post:
```
I think I'd prefer `/deploy` to be a top-level endpoint. For example `curl -X POST https://localhost:17777/deploy ...`.

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        required for uploading the validated resource will be returned.
+      operationId: addApplication
+      parameters:
+        - $ref: "#/components/parameters/ApplicationName"
```
This should be in the `requestBody` I think

cderici replied:
I may be misunderstanding, but I thought this was the identifier that we want the application to have after deploy, no? as in  `juju deploy ubuntu -->my-application<--`, cause the name of the charm to be used to deploy is already in the requestBody (see `charmName` in `DeployArgs`).

jnsgruk replied:
Yep, but I think there is still some misunderstanding based on the latest commits. I don't think we should have a `POST /model{model-name}/application/{app-name}/deploy` - I think we should just have `POST /deploy` and have both the charm name and application name in the request body.

cderici replied:
Ah I see, that makes much more sense now 👍

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        - $ref: "#/components/parameters/Force"
+        - $ref: "#/components/parameters/DryRun"
+        - $ref: "#/components/parameters/Trust"
+        - $ref: "#/components/parameters/NumUnits"
```
```suggestion
        - $ref: "#/components/parameters/Scale"
```

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/DeployFromRepositoryArgs'
```
```suggestion
              $ref: '#/components/schemas/DeployArgs'
```

jnsgruk replied:
I think ^ might be neater :)

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        required: true
+      responses:
+        "200":
+          description: Detailed information about the requested application.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/DeployFromRepositoryResult"
```
Use the `EntityCreated` response here - that way the response body is empty, but we'll return a `Location` header that points to the new resource.

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        - $ref: "#/components/parameters/DryRun"
+      responses:
+        "200":
+          $ref: "#/components/responses/ApplicationDestroyed"
```
```suggestion
          $ref: "#/components/responses/EmptySyncResponse"
```

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+          $ref: "#/components/responses/Forbidden"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+  /application/{name}/refresh:
```
I think this should be `/model/{name}/application/{name}/refresh`

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+          $ref: "#/components/responses/Forbidden"
+        "500":
+          $ref: "#/components/responses/InternalServerError"
+  /application/relate:
```
```suggestion
  /integrate
```

 I think that because `integrate` is not necessarily tied to a model, we should bring this up to the top level

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/AddRelationArgs'
```
```suggestion
              $ref: '#/components/schemas/IntegrateArgs'
```

jnsgruk commented on openapi.yaml:
```diff
@@ -45,6 +45,126 @@ tags:
+        required: true
+      responses:
+        "200":
+          $ref: "#/components/responses/RelationCreated"
```
```suggestion
          $ref: "#/components/responses/EntityCreated"
```

jnsgruk commented on openapi.yaml:
```diff
@@ -155,7 +275,7 @@ paths:
             application/json:
               schema:
-                $ref: "#/components/schemas/UserCloudList"
+                $ref: "#/components/schemas/CloudList"
```
Nice catch, thanks :sunglasses:

jnsgruk commented on openapi.yaml:
```diff
@@ -366,7 +486,7 @@ paths:
             application/json:
               schema:
-                $ref: "#/components/schemas/UserModelList"
+                $ref: "#/components/schemas/ModelList"
```
Good catch, again! :slightly_smiling_face:

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
           $ref: "#/components/responses/InternalServerError"
 components:
   schemas:
+    AddRelationArgs:
```
```suggestion
    IntegrateArgs:
```

jnsgruk replied:
This schema feels a bit anaemic to me.

I could imagine an object that looks a bit like this:

```javascript
{
  "provider": {
    "model": "foo-model",
    "application": "bar-app",
    "endpoint": "http"
  },
  "requirer": {
    "model": "baz-model",
    "application": "quz-app",
    "endpoint": "http"
  }
}
```

Where `model` and `application` are mandatory, and `endpoint` is optional.

WDYT?

jnsgruk replied:
This was marked as resolved, but `IntegrateArgs` still seems to have the old schema.

cderici replied:
Sorry I missed this, fixing.. I think the provider/requirer duo makes sense :+1:

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+      properties:
+        base:
+          $ref: '#/components/schemas/Base'
+        channel:
+          type: string
```
I wonder if we should have a `CharmChannel` object that looks like:

```json
{
  "track": "latest",
  "channel": "edge",
  "branch": "pr-45"
}
```

WDYT?

cderici replied:
Makes sense, but with what you probably meant `"risk": "edge"` (instead of `"channel":  "edge"`). Also are the branches still in use?

jnsgruk replied:
Yes, you're right that it should probably be:

```javascript
{
  "track": "latest",
  "risk": "edge",
  "branch": "pr-45"
}
```

And yes, branches are still in use! :)

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          $ref: '#/components/schemas/Constraints'
+        endpoint-bindings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
```
Can we remove this `patternProperties` by either: modelling endpoint bindings as a list of well-defined objects, or more simply using the suggestion here.

```suggestion
          additionalProperties: true
```

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: boolean
+        exposed-endpoints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/ExposedEndpoint'
```
As above.

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/ExposedEndpoint'
+        life:
+          type: string
```
What is this?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: string
+        principal:
+          type: boolean
+        remote:
+          type: boolean
```
What is this?

cderici replied:
That's a flag indicating if this application is a RemoteApplication involved in a cross-model relation. I always find this flag weird, because the application is "remote" only from the other side's point of view. We can take it out for now and bring it back when thinking about cross-model relations. WDYT?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: boolean
+        remote:
+          type: boolean
+        tag:
+          type: string
```
Should we remove this?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: boolean
+        tag:
+          type: string
+      additionalProperties: false
```
Should there be a `revision` field in here?

cderici replied:
No I think we should have a `CharmOrigin` here (and that should have that info in it).

jnsgruk replied:
Ah in that case we can probably drop `base`, `channel` and other fields that are contained within `CharmOrigin`

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          $ref: '#/components/schemas/CharmOrigin'
+        charm-url:
+          type: string
+        config-settings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        config-settings-yaml:
+          type: string
+        endpoint-bindings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        force:
+          type: boolean
+        force-base:
+          type: boolean
+        force-units:
+          type: boolean
+        generation:
+          type: string
```
I think these should be removed -- `force`, `force-base`, etc. feel like URL query parameters to me. Config should be handled using a separate endpoint

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: string
+        resource-ids:
+          type: object
+          patternProperties:
+            .*:
+              type: string
```
This feels like it needs more modelling - should this be a list of IDs? a list of resource objects?

cderici replied:
That's supposed to be a map of resource names to resource IDs, right? i.e., `map[string]string`, so I think the following should model it just fine, wdyt?:

```yaml
type: object
additonalProperties:
  type: string
```

jnsgruk replied:
Perhaps it could be a list of `Resource` objects, where `Resource` has fields `name`, `id` and `type` (type is an `enum` containing `file` and `oci-image`)?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          patternProperties:
+            .*:
+              type: string
+        storage-constraints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/StorageConstraints'
```
I'm not sure this belongs here?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+        - exposed
+        - remote
+        - life
+    ApplicationSetCharm:
```
I think this should probably be called something else (can't quite decide what, yet) and only include:

- `charm`: the name of the charm package (as per metadata.yaml)
- `channel`: a well defined CharmChannel as suggested above
- `origin`: a `CharmOrigin` object as it is here
- `url`: as defined here (I assume this points to a charmhub uri?)

cderici replied:
I changed its name to `ApplicationSubstitute` for now as a suggestion. It can always be changed and improved.

So currently in Juju the charm-url would have the charm name, unless it's a local charm, in which case it's something like `local:<path>`. But we really want to eventually get rid of the charm URL (we have been discussing that since Prague). As long as we have the origin (that has the source --charmhub/local--, base, track and the risk), we shouldn't have any trouble locating the charm. 

If the client is refreshing with a local charm, then it'll send the charm in the request body anyways, right? That part might need some modelling.

> and only include:

Are you suggesting we should not support things like resetting configs, e.g. `juju refresh foo --config config.yaml`, or uploading resources, e.g. `juju refresh foo --resource bar=/some/file.tgz` or `juju refresh foo --storage cache=ssd,10G`? Or are you just saying those can be added later once the main structure is established?

jnsgruk replied:
Ok so happy with `ApplicationSubstitute` for now, though if we're including the `CharmOrigin` object as part of it then there are some fields that can be removed (`base`, `arch`, `track`, etc.) from the `ApplicationSubsitute`

You're right about `juju refresh foo --config ...` etc., we probably do want to allow for that.

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+        - force
+        - force-units
+        - force-base
+    Base:
+      type: object
+      properties:
+        channel:
+          type: string
+        name:
+          type: string
+      additionalProperties: false
+      required:
+        - name
+        - channel
```
Is this required? I wonder if we can come up with slightly better name or modelling for this?

cderici replied:
I'm not entirely sure what you're referring to, are we talking about `Base`, or you mean the `channel` there?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+      required:
+        - name
+        - channel
+    CharmRelation:
```
```suggestion
    CharmIntegration:
```

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: string
+        optional:
+          type: boolean
+        role:
+          type: string
```
Make this an enum and list the possible values (see `ModelUserAccess` for an example)

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: boolean
+        role:
+          type: string
+        scope:
+          type: string
```
Make this an enum as above I think

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+      type: object
+      properties:
+        architecture:
+          type: string
```
Make this a commonly defined schema object in itself with an enum? I expect we'll use this in lots of places.

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+      properties:
+        architecture:
+          type: string
+        base:
+          $ref: '#/components/schemas/Base'
```
I think if we remove this as I suggested above, then we might just need a `channel` field in here that is a `CharmChannel` object containing track/channel/branch?

cderici replied:
So it's important to make a terminology distinction here. 

In Juju, we have a [series.Channel](https://github.com/juju/juju/blob/2b2308500d71d17867256cc81dce9dbce877bcc4/core/series/channel.go#L40) that the [Base](https://github.com/juju/juju/blob/2b2308500d71d17867256cc81dce9dbce877bcc4/core/series/base.go#L15) uses, which looks something like `22.04@stable` (track/risk -- doesn't have branch). And we also have a [charm.Channel](https://github.com/juju/charm/blob/6beed86562dc206e5e53aeb168eb94aeaf4d6729/channel.go#L40) that looks like  `latest/stable/br`, which was the prominent "channel" in the past. Lately we have been moving away from using the charm.Channel and using mostly the Base (which also has the OS info). So I suggest sticking with that. At the very least, I think we should refrain from using two different `channel`s, and pick just one of them. Otherwise we'll inherit this nasty channel overload.

jnsgruk replied:
Well it's not necessarily that easy I don't think - for a given application deployment you have **both** a base channel, and a charm channel at play?

cderici replied:
Oh you're right, disregard my comment above, the channel information (for both types of channels) has to be there. Though I will suggest not to call both of them `channel`s.

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          $ref: '#/components/schemas/Base'
+        branch:
+          type: string
+        hash:
+          type: string
```
Hash of what?

cderici replied:
I think that's the hash of the charm in the charmhub located by this origin, presumably will only be set in the response of `GET /model{model-name}/application/{app-name}`

jnsgruk replied:
Can you add a description to that field (just because it's not obvious)?

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: string
+        id:
+          type: string
+        instance-key:
+          type: string
```
What is this?

cderici replied:
Apparently it is a unique string associated with the application.

https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/rpc/params/applications.go#L42

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: integer
+        risk:
+          type: string
+        source:
+          type: string
```
Is this a URI or something else? I'm not sure what this field is?

cderici replied:
https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/core/charm/origin.go#L16C46-L16C46

jnsgruk replied:
So that's a URI? Probably worth adding a description to this field, or at least a comment for now. Perhaps include an example?

cderici replied:
Oh I'm sorry the link was a bit ambiguous, it's just an enum of ["local", "charmhub"]. [This](https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/core/charm/origin.go#L30) would be a better spot to show.

jnsgruk replied:
Great, let's get that modelled as an enum then :)

cderici replied:
Already done 👍

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +591,173 @@ paths:
+          type: string
+        track:
+          type: string
+        type:
+          type: string
```
What does this denote? `principal` or `subordinate` or something else?

cderici replied:
That's if it's for a `charm`  or a ` bundle` . I'll make it a simple enum.

https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/charmhub/transport/common.go#L21

https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/cmd/juju/application/store/charmadapter.go#L88C2-L88C2

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+        allocate-public-ip:
+          type: boolean
+        arch:
+          type: string
```
This should be a common field that's represented as an enum (as above)

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+        arch:
+          type: string
+        container:
+          type: string
```
Make this an enum? Perhaps a defined object itself if we're using in multiple places.

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
       additionalProperties: false
       required:
         - permissions
+    Constraints:
+      type: object
+      properties:
+        allocate-public-ip:
+          type: boolean
+        arch:
+          type: string
+        container:
+          type: string
+        cores:
+          type: integer
+        cpu-power:
+          type: integer
```
What is this?

cderici replied:
Most of these are pretty technical, adding descriptions and modelling some of them now :+1: 

https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/core/constraints/constraints.go#L59

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+        image-id:
+          type: string
+        instance-role:
+          type: string
```
What is this? Is this per-cloud?

cderici replied:
I'm not sure about this, @hmlanigan any ideas?

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
       additionalProperties: false
       required:
         - permissions
+    Constraints:
+      type: object
+      properties:
+        allocate-public-ip:
+          type: boolean
+        arch:
+          type: string
+        container:
+          type: string
+        cores:
+          type: integer
+        cpu-power:
+          type: integer
+        image-id:
+          type: string
+        instance-role:
+          type: string
+        instance-type:
+          type: string
+        mem:
+          type: integer
+        root-disk:
+          type: integer
+        root-disk-source:
+          type: string
+        spaces:
+          type: array
+          items:
+            type: string
```
Should this be an array of **strings**? Is it just a list of space IDs?

cderici replied:
I'm not super familiar with spaces, here's a description I found in [here](https://github.com/juju/juju/blob/c0ab5b627ffa175f439bb2357c33621fd5353e1a/core/constraints/constraints.go#L95):

```
        // holds a list of juju network spaces that
	// should be available (or not) on the machine. Positive and
	// negative values are accepted, and the difference is the latter
	// have a "^" prefix to the name.
```

Adding description for it.

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+          items:
+            type: string
+      additionalProperties: false
+    DeployFromRepositoryArgs:
```
```suggestion
    DeployArgs:
```

jnsgruk replied:
Some of the keys in this object start with upper case, some are lower case. Can we please standardise on lower case?

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+          type: string
+        ConfigYAML:
+          type: string
+        Cons:
```
just make this `constraints`

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+          type: string
+        Cons:
+          $ref: '#/components/schemas/Constraints'
+        Devices:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/Constraints'
```
This probably needs modelling slightly better

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/Constraints'
```
This needs modelling properly

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+      required:
+        - charm-url
+        - architecture
+    DeployFromRepositoryResult:
+      type: object
+      properties:
+        Info:
+          $ref: '#/components/schemas/DeployFromRepositoryInfo'
+        PendingResourceUploads:
+          type: array
+          items:
+            $ref: '#/components/schemas/PendingResourceUpload'
+      additionalProperties: false
+      required:
+        - Info
+        - PendingResourceUploads
```
```suggestion
```

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+    DeployFromRepositoryInfo:
+      type: object
+      properties:
+        architecture:
```
I notice that in some places you call this `arch` and others `architecture`. Let's standardise on `architecture`?

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+          $ref: '#/components/schemas/Base'
+        channel:
+          type: string
+        endpoint-bindings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
```
Avoid patternProperties please :)

jnsgruk commented on openapi.yaml:
```diff
@@ -626,6 +913,140 @@ components:
+      type: object
+      description: Information about the local resources that need to be uploaded by the client after a deploy operation.
+      properties:
+        Filename:
+          type: string
+        Name:
+          type: string
+        Type:
+          type: string
+        pending-id:
+          type: string
```
Inconsistent capitalisation here

jnsgruk commented on openapi.yaml:
```diff
@@ -1037,6 +1470,17 @@ components:
+        directive:
+          type: string
+        scope:
+          type: string
```
I don't know enough about this right now, but I wonder if this could be modelled more robustly so it's not just arbitrary strings?

jnsgruk commented on openapi.yaml:
```diff
@@ -1135,12 +1609,20 @@ components:
+      in: query
+      name: dry-run
+      required: false
+      schema:
+        type: boolean
+        default: false
```
```suggestion
      schema:
        type: string
        default: false
        enum:
          - true
          - false
```

I now realise I've made this mistake in some of the other query parameters too, but I don't think query params work well with the `boolean` type, and it's better to use a string enum with just true/false.

jnsgruk commented on openapi.yaml:
```diff
@@ -1154,12 +1636,25 @@ components:
       description: The name of a model in the controller.
       schema:
         type: string
+    NumUnits:
```
```suggestion
    Scale:
```

jnsgruk commented on openapi.yaml:
```diff
@@ -1168,6 +1663,22 @@ components:
       schema:
         type: string
   responses:
+    ApplicationDestroyed:
+      type: object
+      properties:
+        destroyed-storage:
+          type: array
+          items:
+            type: string
+        destroyed-units:
+          type: array
+          items:
+            type: string
+        detached-storage:
+          type: array
+          items:
+            type: string
+      additionalProperties: false
```
```suggestion
```

jnsgruk commented on openapi.yaml:
```diff
@@ -1222,6 +1715,12 @@ components:
         application/json:
           schema:
             $ref: '#/components/schemas/Error'
+    RelationCreated:
```
```suggestion
    IntegrationCreated:
```

jnsgruk commented on openapi.yaml:
```diff
@@ -1222,6 +1715,12 @@ components:
           schema:
             $ref: '#/components/schemas/Error'
+    RelationCreated:
+      description: Relation added between specified endpoints.
```
```suggestion
      description: Integration added between specified endpoints.
```

jnsgruk commented on openapi.yaml:
```diff
@@ -1222,6 +1715,12 @@ components:
+      content:
+        application/json:
+          schema:
+            $ref: '#/components/schemas/CharmRelation'
```
```suggestion
            $ref: '#/components/schemas/CharmIntegration'
```

jnsgruk reviewed (changes requested):
Thank you for having a go at this @cderici. It probably looks like LOADS of comments here, but overall the theme is: let's try to model the API and schemas that we *want*, not necessarily stick with what we already *have*.

The current facades are quite fractured, in that we have lots of data structures that represent the same object with slightly varying fields, and I'd like to get away from that if possible. 

Take a look through this, and perhaps we could jump on a call and work through in a bit more detail if that's useful?

jnsgruk reviewed (changes requested):
Thank you for having a go at this @cderici. It probably looks like LOADS of comments here, but overall the theme is: let's try to model the API and schemas that we *want*, not necessarily stick with what we already *have*.

The current facades are quite fractured, in that we have lots of data structures that represent the same object with slightly varying fields, and I'd like to get away from that if possible. 

Take a look through this, and perhaps we could jump on a call and work through in a bit more detail if that's useful?

cderici committed:
Add application tag

cderici committed:
arch -> architecture

cderici committed:
Replace response with EntityCreated

cderici committed:
Standardize capitalization

cderici committed:
Apply suggestions from code review

Co-authored-by: Jon Seager <jon@sgrs.uk>
Signed-off-by: Caner Derici <caner.derici@canonical.com>

cderici committed:
Use model name in identifying applications

Also make the parameter names explicit

cderici committed:
Make deploy a toplevel endpoint

cderici committed:
Use model name in application.refreh to id the app

cderici committed:
Model CharmChannel as an object

cderici committed:
Remove the patternProperties

cderici committed:
Tidy up Application a little bit

cderici committed:
Tidy up application refresh based on review

cderici committed:
Add enums of possible values in CharmIntegration

cderici committed:
Add Architecture object as an enum of possible arch

cderici committed:
Revisit CharmOrigin

jnsgruk commented on openapi.yaml:
```diff
@@ -471,6 +598,172 @@ paths:
+        - principal
+        - exposed
+        - remote
+        - life
```
```suggestion
```

This field isn't included in the object any more

jnsgruk commented:
Left a few more comments to follow up on progress so far! Thanks for you work here, and let me know when you want another full pass!

jnsgruk committed:
feat: add user endpoints

jnsgruk committed:
Merge branch 'main' into some-application-endpoints

Signed-off-by: Jon Seager <jon@sgrs.uk>

cderici commented:
> Left a few more comments to follow up on progress so far! Thanks for you work here, and let me know when you want another full pass!

Sounds good, still got a couple of items to go through (doing it in the spare time in the evenings so it's going a bit slower during the week :), I'll let you know when I'm done with all your comments. Maybe we can do the next pass together on a call, it might help me get even more in line with the design in your mind 👍

jnsgruk committed:
Merge branch 'main' into some-application-endpoints

Signed-off-by: Jon Seager <jon@sgrs.uk>

jnsgruk commented:
Sounds good to me -- I've just resolved a couple of conflicts from where I've been hacking on it today, so make sure to do a quick pull before you restart!

jnsgruk committed:
Merge branch 'main' into some-application-endpoints

Signed-off-by: Jon Seager <jon@sgrs.uk>

jnsgruk committed:
Merge branch 'main' into some-application-endpoints

Signed-off-by: Jon Seager <jon@sgrs.uk>

jnsgruk committed:
fix: correct some linting errors

cderici committed:
Optimize fields in Application and CharmOrigin

cderici committed:
Add description for CharmOrigin.hash

cderici committed:
Make Constraints clearer

cderici committed:
fix: model Placement.scope and add descriptions

cderici committed:
fix: remove non-existent field from required

Co-authored-by: Jon Seager <jon@sgrs.uk>
Signed-off-by: Caner Derici <caner.derici@canonical.com>

cderici committed:
fix: endpoints in readme

cderici committed:
fix: add description for CharmOrigin.instance-key

cderici committed:
fix: add default parameters in refresh endpoint

cderici committed:
fix: remove redundant channel from ApplicationSubstitute

cderici commented:
@jnsgruk I think we're good to go for a whole pass on this. We should probably have a chat about the `PendingResourceUploads` that needs to be returned from the deploy, but other than that everything seems clear to me if nothing slipped through the cracks 👍

jnsgruk commented:
Have booked some time tomorrow - still looks like there are some unresolved conversations further up - would be good if you manage to take a look before we meet, but otherwise we'll cover them then!

I also introduced a linter you can run with `make lint` which is flagging the following:

```
❯ make lint
npx @stoplight/spectral-cli lint /home/jon/data/code/canonical/juju-api/openapi.yaml

/home/jon/data/code/canonical/juju-api/openapi.yaml
 1112:30  warning  oas3-unused-component  Potentially unused component has been detected.  components.schemas.DeployFromRepositoryInfo
 1129:27  warning  oas3-unused-component  Potentially unused component has been detected.  components.schemas.PendingResourceUpload
 1615:24  warning  oas3-unused-component  Potentially unused component has been detected.  components.schemas.StorageConstraints
 1862:24  warning  oas3-unused-component  Potentially unused component has been detected.  components.responses.IntegrationCreated

✖ 4 problems (0 errors, 4 warnings, 0 infos, 0 hints)
```

jnsgruk committed:
Merge branch 'main' into some-application-endpoints

cderici commented:
> still looks like there are some unresolved conversations further up

@jnsgruk Oh looks like they don't appear at the `Files changed` tab, which is why I didn't realize, I'll take a look at them today 👍 

> I also introduced a linter you can run with `make lint` which is flagging the following:

Not only that, looks like it's also putting comments in the changes, very cool!

![image](https://github.com/jnsgruk/juju-api/assets/120652/c99a0cc1-ad51-43ce-9da2-32286cc21b83)

cderici committed:
fix: revise ApplicationSubstitute fields based on review

cderici committed:
fix: remove unused components

keeping PendingResourceUpload for now for the discussion about
deploying charms with local resources

cderici committed:
fix: bring deploy endpoint to toplevel

cderici committed:
feat: add schema for Resource

cderici committed:
feat: add Integrator for IntegrateArgs

jnsgruk commented on openapi.yaml:
```diff
@@ -722,6 +981,131 @@ components:
+    DeployArgs:
+      type: object
+      properties:
+        model-name:
```
```suggestion
        model:
```

jnsgruk commented on openapi.yaml:
```diff
@@ -722,6 +981,131 @@ components:
+          type: string
+        application-name:
+          type: string
+        attachStorage:
```
camel case or hyphen seperated?

hmlanigan replied:
hyphen would match the charm hub api

jnsgruk commented on openapi.yaml:
```diff
@@ -722,6 +981,131 @@ components:
+            type: string
+        charmName:
+          type: string
+        configYAML:
```
camel case or hyphenated?

jnsgruk commented on openapi.yaml:
```diff
@@ -722,6 +981,131 @@ components:
+        revision:
+          type: integer
+      additionalProperties: false
+      required:
+        - charmName
```
model should be in here?

jnsgruk commented on openapi.yaml:
```diff
@@ -722,6 +981,131 @@ components:
       additionalProperties: false
       required:
         - permissions
+    Constraints:
```
This needs breaking down into "storage constraints" "device constraints" etc.

cderici replied:
I'm not sure what to do here. The only fields in the `Constraints` that could be included in the `StorageConstraints` (which we already have) are the `root-disk` and the `root-disk-source`, which looking at their description seem awfully related to the device. 

Maybe I can just rename this `Constraints -> DeviceConstraints` or `Constraints -> UnitHardwareConstraints` (which might be more descriptive)? WDYT?

jnsgruk commented on openapi.yaml:
```diff
@@ -567,6 +692,140 @@ paths:
+      type: object
+      description: Information about a given application.
+      properties:
+        charm-origin:
+          $ref: '#/components/schemas/CharmOrigin'
```
```suggestion
```

jnsgruk replied:
Base, channel and revision should be added

hmlanigan replied:
```
"base":{"architecture":"amd64","name":"ubuntu","channel":"22.04"}}]
```
From the charm hub api, though we don't need a list. The current juju api defines a base as a name/channel. Do we where do we want to include the architecture? Today is in constraints.

cderici replied:
Currently our base looks like following:

```yaml
    Base:
      type: object
      properties:
        channel:
          type: string
        name:
          type: string
      additionalProperties: false
      required:
        - name
        - channel
```

Made the changes requested above.

jnsgruk commented on openapi.yaml:
```diff
@@ -567,6 +692,140 @@ paths:
+            $ref: '#/components/schemas/ExposedEndpoint'
+        principal:
+          type: boolean
+      additionalProperties: false
```
There should be a list of resources returned here so that the deploy request can be followed by another to upload resources if required

POST /models/foo/applications/postgresql/resources/massive-file

hmlanigan commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Annotations          | Set                             |                                                                                              |
+| Application          | AddRelation                     | `POST /integrate`                                                                            |
+| Application          | AddUnits                        |                                                                                              |
+| Application          | ApplicationsInfo                | `GET /model{model}/application/{application}`                                                |
```
```suggestion
| Application          | ApplicationsInfo                | `GET /model/{model}/application/{application}`                                                |
```

hmlanigan commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Application          | MergeBindings                   |                                                                                              |
+| Application          | ResolveUnitErrors               |                                                                                              |
+| Application          | ScaleApplications               |                                                                                              |
+| Application          | SetCharm                        | `PATCH /model/{model}/application/{application}/refresh?force=false&dry-run=false`           |
```
I wonder if we need SetCharm, it'd used by deploy and refresh, but if we get those correct. May not be needed in the new API.

cderici replied:
Not sure exactly what you mean. Do you mean like to make the `/refresh` a toplevel endpoint?

hmlanigan commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Application          | Consume                         |                                                                                              |
+| Application          | Deploy                          |                                                                                              |
+| Application          | DeployFromRepository            | `POST /model/{model}/application/{application}/deploy?force=false&dry-run=false&trust=false` |
+| Application          | DestroyApplication              | `DELETE /model/{model}/application/{application}`                                            |
```
```suggestion
| Application          | DestroyApplication              | `DELETE /model/{model}/application/{application}?force=false`                                            |
```

hmlanigan commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Application          | CharmRelations                  |                                                                                              |
+| Application          | Consume                         |                                                                                              |
+| Application          | Deploy                          |                                                                                              |
+| Application          | DeployFromRepository            | `POST /model/{model}/application/{application}/deploy?force=false&dry-run=false&trust=false` |
```
```suggestion
| Application          | DeployFromRepository            | `POST /model/{model}/application/{application}/deploy&dry-run=false&trust=false` |
```

We've been talking about removing force from deploy. It was used for testing purposes previously and for un-updated charms. Since we now have Branches as part of charm channels, there is a way to test. As well, with multiple architectures forcing a charm deploy is more risky

hmlanigan commented on openapi.yaml:
```diff
@@ -305,6 +362,75 @@ paths:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ApplicationSubstitute'
```
What does ApplicationSubstitute stand for?

cderici replied:
It's a model of an application that the user wants to replace with a current one. (i.e., `ApplicationSetCharm` in the current juju api).

hmlanigan commented on openapi.yaml:
```diff
@@ -567,6 +692,140 @@ paths:
+      additionalProperties: false
+      required:
+        - application
+        - charm-url
```
charm name, we should remove any reference to charm-url in this doc

hmlanigan commented on openapi.yaml:
```diff
@@ -567,6 +692,140 @@ paths:
+      required:
+        - application
+        - charm-url
+        - charm-origin
```
not required.

hmlanigan commented on openapi.yaml:
```diff
@@ -1133,6 +1542,22 @@ components:
+          description: Scope-specific placement directive.
+          type: string
+        scope:
+          description: Can be a container type or a machine scope (#).
```
You can specify a few provider specific directives also, such as availability zone, all though a zone can be a constraint as well. Check for others which are not constraints.  Some placement directives are specific to bundle and won't be used here.

cderici replied:
I think you mean the [environment names ](https://github.com/juju/juju/blob/30ba85cdcc5868c64fb9dd9c3f50330d1fa5fac1/core/instance/placement.go#L24), which is not included here (i.e. only containers and existing machine IDs are accepted). Constraint-like things such as the zones are modeled in the `Constraints` schema.

hmlanigan commented on openapi.yaml:
```diff
@@ -1279,12 +1746,23 @@ components:
+      name: dry-run
+      required: false
+      schema:
+        type: string
```
```suggestion
        type: boolean
```

cderici committed:
Apply suggestions from code review

Co-authored-by: hmlanigan <heather.lanigan@canonical.com>
Co-authored-by: Jon Seager <jon@sgrs.uk>
Signed-off-by: Caner Derici <caner.derici@canonical.com>

cderici committed:
fix: replace origin with base/channel/revision in Application

cderici committed:
fix: camelCase -> hypened

cderici committed:
fix: adjust required fields

cderici committed:
feat: add resources field to Application response

cderici committed:
feat: add endpoint to upload resources post-deploy

cderici commented:
@jnsgruk I think this is getting ready to land, unless I'm missing anything. I did everything we talked about, except a couple of parts I didn't fully understand (kept those convos unresolved). Can you go over this whenever you get the chance? The PR is starting to get a bit large/hard to control with 122 conversations - I keep missing to address stuff every time :), so it might be better to continue with smaller changes. WDYT?

jnsgruk commented:
Thanks @cderici - agreed. I'll take a look sometime this week with a view to merging this :rocket:

jnsgruk commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Application          | CharmConfig                     |                                                                                              |
+| Application          | CharmRelations                  |                                                                                              |
+| Application          | Consume                         |                                                                                              |
+| Application          | Deploy                          |                                                                                              |
+| Application          | DeployFromRepository            | `POST /model/{model}/application/{application}/deploy&dry-run=false&trust=false` |
```
@cderici this needs updating to reflect the changes in the actual openapi spec :)

jnsgruk commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Application          | Consume                         |                                                                                              |
+| Application          | Deploy                          |                                                                                              |
+| Application          | DeployFromRepository            | `POST /model/{model}/application/{application}/deploy&dry-run=false&trust=false` |
+| Application          | DestroyApplication              | `DELETE /model/{model}/application/{application}?force=false`                                            |
```
```suggestion
| Application          | DestroyApplication              | `DELETE /model/{model}/application/{application}`                                            |
```

jnsgruk replied:
Let's just ensure the force is the default and leave this out here :)

jnsgruk commented on README.md:
```diff
@@ -51,219 +51,219 @@ go run schemagen.go -admin-facades -facade-group client ./schemas/client-schemas
+| Application          | MergeBindings                   |                                                                                              |
+| Application          | ResolveUnitErrors               |                                                                                              |
+| Application          | ScaleApplications               |                                                                                              |
+| Application          | SetCharm                        | `PATCH /model/{model}/application/{application}/refresh?force=false&dry-run=false`           |
```
```suggestion
| Application          | SetCharm                        | `PATCH /model/{model}/application/{application}/refresh`           |
```

jnsgruk commented on openapi.yaml:
```diff
@@ -50,6 +55,58 @@ tags:
+    post:
+      tags:
+        - application
+      description: Deploy an application. If a local resource is provided, details
+        required for uploading the validated resource will be returned.
+      # This needs to return info about local resources to be uploaded by the client
+      # post-deploy (PendingResourceUpload)
```
```suggestion
      description: |
        Deploy an application. 

        Returns a simple response with a location header pointing to the deployed
        application.

        If the deployment relies upon local resources being uploaded, clients should
        retrieve the URL in the location header send in the response in order to
        learn the URL resources should be POSTed too after deploy (see `uploadResource`
        operation).
```

Note the `description: |` part here too -- can you please ensure that's tidied up anywhere you've got a newline in a description?

jnsgruk commented on openapi.yaml:
```diff
@@ -305,6 +362,100 @@ paths:
+      tags:
+        - application
+      description: Uploads a resource for an application
+      operationId: addResource
```
```suggestion
      operationId: uploadResource
```

jnsgruk commented on openapi.yaml:
```diff
@@ -739,11 +1137,36 @@ components:
         - error_code
         - type
-    MachineHardware:
+    Intergrator:
```
Typo! `s/Intergrator/Integrator/g` :wink:

jnsgruk reviewed (changes requested):
Thank you for all this work!! Starting to look really good :muscle: 

Some final adjustments to spelling/wording, and then we're good to merge. Gimme a ping when you've got through them. :rocket:

cderici committed:
Apply suggestions from code review

Co-authored-by: Jon Seager <jon@sgrs.uk>
Signed-off-by: Caner Derici <caner.derici@canonical.com>

cderici committed:
fix: endpoint for deploy in readme

cderici committed:
fix: typo integrator

cderici committed:
fix: revisit descriptions

cderici committed:
fix: remove redundant newline

cderici commented:
@jnsgruk added the changes you mentioned, I think we're good to go :+1:
